cmd/gcp-controller-manager/readonly_resource_csr_approver.go

```diff
@@ -0,0 +1,310 @@
+}
+
+func (approver *readonlyResourceApprover) updateCSR(csr *capi.CertificateSigningRequest, approved bool, msg string) error {
+	if approved {
+		csr.Status.Conditions = append(csr.Status.Conditions, capi.CertificateSigningRequestCondition{
+			Type:    capi.CertificateApproved,
+			Reason:  "AutoApproved",
+			Message: msg,
+			Status:  v1.ConditionTrue,
+		})
+	} else {
+		csr.Status.Conditions = append(csr.Status.Conditions, capi.CertificateSigningRequestCondition{
+			Type:    capi.CertificateDenied,
+			Reason:  "AutoDenied",
+			Message: msg,
+			Status:  v1.ConditionTrue,
+		})
+	}
+	updateRecordMetric := csrmetrics.OutboundRPCStartRecorder("k8s.CertificateSigningRequests.updateApproval")
+	_, err := approver.ctx.client.CertificatesV1().CertificateSigningRequests().UpdateApproval(context.TODO(), csr.Name, csr, metav1.UpdateOptions{})
+	if err != nil {
+		klog.Errorf("error updating approval status for csr: %v", err)
+		updateRecordMetric(csrmetrics.OutboundRPCStatusError)
+		return fmt.Errorf("error updating approval status for csr: %v", err)
+	}
+	updateRecordMetric(csrmetrics.OutboundRPCStatusOK)
+	return nil
+}
+
+type readonlyResourceCsrValidator struct {
+	name          string
+	authFlowLabel string
+
+	// validate will validate the request and return approved or denied
+	// (false, err) will base on err type retry or not. if it is a non-retryable error then deny the request
+	// (true, nil) will continue to next validater. if it is the last one, will approve the request
+	// (false, nil) reject the request.
+	validate   readonlyResourceCSRValidateFunc
+	permission authorization.ResourceAttributes
+}
+
+func newReadonlyResourceCsrValidator() []readonlyResourceCsrValidator {
+	return []readonlyResourceCsrValidator{
+		{
+			name:          "csr usage validator",
+			authFlowLabel: "readonly_resource_csr_usage_match",
+			validate:      validateCSRUsage,
+		},
+		{
+			name:          "pod annotation validator",
+			authFlowLabel: "readonly_resource_pod_annotation_valid",
+			validate:      validatePodAnnotation,
+		},
+		{
+			name:          "rbac validator",
+			authFlowLabel: "readonly_resource_rbac_exist",
+			validate:      validateRbac,
+		},
+	}
+}
+
+func validateCSRUsage(content readonlyResourceCSRRequestContent) readonlyResourceCSRRequestResult {
+	csr := content.csr
+	if csr == nil || len(csr.Spec.Usages) == 0 {
+		klog.Errorf("csr usage is empty")
+		return readonlyResourceCSRRequestResult{
+			result:  false,
+			err:     nil,
+			message: "csr usage is empty",
+		}
+	}
+
+	usageMap := map[capi.KeyUsage]struct{}{}
+	for _, u := range kubeletClientUsages {
+		usageMap[u] = struct{}{}
+	}
+
+	for _, usage := range csr.Spec.Usages {
+		if _, ok := usageMap[usage]; !ok {
+			klog.Errorf("csr usage %v is not allowed", usage)
+			return readonlyResourceCSRRequestResult{
+				result:  false,
+				err:     nil,
+				message: fmt.Sprintf("csr usage %v is not allowed", usage),
```

Comment:
OK, just a little concern, it may provide more information to attacker, but since we have publish the code on github, it does no harm to inform the user, will do